File: server/src/main/java/com/cupboard/api/enums/Currency.java
package com.cupboard.api.enums;

public enum Currency {
    USD("US Dollar", 100),
    EUR("Euro", 100),
    GBP("British Pound", 100),
    JPY("Japanese Yen", 1),
    KRW("South Korean Won", 1),
    VND("Vietnamese Dong", 1),
    CAD("Canadian Dollar", 100),
    AUD("Australian Dollar", 100),
    NZD("New Zealand Dollar", 100),
    HKD("Hong Kong Dollar", 100),
    SGD("Singapore Dollar", 100),
    PHP("Philippine Peso", 100),
    THB("Thai Baht", 100),
    CNY("Chinese Yuan", 100);

    private final String displayName;
    private final int subunitDivisor;

    Currency(String displayName, int subunitDivisor) {
        this.displayName = displayName;
        this.subunitDivisor = subunitDivisor;
    }

    public String getDisplayName() { return displayName; }
    public int getSubunitDivisor() { return subunitDivisor; }

    public double toDisplayAmount(long smallestUnit) {
        return (double) smallestUnit / subunitDivisor;
    }
}

File: client/src/lib/currency.ts
export const CURRENCIES = {
  USD: { displayName: 'US Dollar',          subunitDivisor: 100, symbol: '$'   },
  EUR: { displayName: 'Euro',               subunitDivisor: 100, symbol: '€'   },
  GBP: { displayName: 'British Pound',      subunitDivisor: 100, symbol: '£'   },
  JPY: { displayName: 'Japanese Yen',       subunitDivisor: 1,   symbol: '¥'   },
  KRW: { displayName: 'South Korean Won',   subunitDivisor: 1,   symbol: '₩'   },
  VND: { displayName: 'Vietnamese Dong',    subunitDivisor: 1,   symbol: '₫'   },
  CAD: { displayName: 'Canadian Dollar',    subunitDivisor: 100, symbol: 'CA$' },
  AUD: { displayName: 'Australian Dollar',  subunitDivisor: 100, symbol: 'A$'  },
  NZD: { displayName: 'New Zealand Dollar', subunitDivisor: 100, symbol: 'NZ$' },
  HKD: { displayName: 'Hong Kong Dollar',   subunitDivisor: 100, symbol: 'HK$' },
  SGD: { displayName: 'Singapore Dollar',   subunitDivisor: 100, symbol: 'S$'  },
  PHP: { displayName: 'Philippine Peso',    subunitDivisor: 100, symbol: '₱'   },
  THB: { displayName: 'Thai Baht',          subunitDivisor: 100, symbol: '฿'   },
  CNY: { displayName: 'Chinese Yuan',       subunitDivisor: 100, symbol: '¥'   },
} as const

export type CurrencyCode = keyof typeof CURRENCIES

export function formatCurrency(amount: number, currency: CurrencyCode): string {
  const { subunitDivisor } = CURRENCIES[currency]
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
  }).format(amount / subunitDivisor)
}

export function toSmallestUnit(amount: number, currency: CurrencyCode): number {
  return Math.round(amount * CURRENCIES[currency].subunitDivisor)
}

export function fromSmallestUnit(amount: number, currency: CurrencyCode): number {
  return amount / CURRENCIES[currency].subunitDivisor
}

export const CURRENCY_OPTIONS = Object.entries(CURRENCIES).map(([code, val]) => ({
  value: code as CurrencyCode,
  label: `${code} — ${val.displayName}`,
}))
